File: internal/domain/progression_test.go
package domain

import "testing"

func TestProgressionStrategy_IsValid(t *testing.T) {
	valid := []ProgressionStrategy{ProgressionLinear, ProgressionPercentage, ProgressionWave}
	for _, s := range valid {
		if !s.IsValid() {
			t.Errorf("expected %q to be valid", s)
		}
	}
	invalid := []ProgressionStrategy{"", "invalid", "LINEAR", "none"}
	for _, s := range invalid {
		if s.IsValid() {
			t.Errorf("expected %q to be invalid", s)
		}
	}
}

func TestProgressionRule_Validate(t *testing.T) {
	inc5 := 5.0
	pct2 := 2.5
	zero := 0.0
	neg := -1.0

	tests := []struct {
		name     string
		r        ProgressionRule
		wantErr  bool
		errField string
	}{
		{
			name: "valid linear",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &inc5,
				DeloadThreshold: 3, DeloadPct: 10,
			},
		},
		{
			name: "valid percentage",
			r: ProgressionRule{
				Strategy: ProgressionPercentage, IncrementPct: &pct2,
				DeloadThreshold: 3, DeloadPct: 10,
			},
		},
		{
			name: "valid wave",
			r: ProgressionRule{
				Strategy:        ProgressionWave,
				DeloadThreshold: 3, DeloadPct: 10,
			},
		},
		{
			name: "invalid strategy",
			r: ProgressionRule{
				Strategy: "invalid", DeloadThreshold: 3, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "strategy",
		},
		{
			name: "linear missing increment",
			r: ProgressionRule{
				Strategy:        ProgressionLinear,
				DeloadThreshold: 3, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "increment",
		},
		{
			name: "linear zero increment",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &zero,
				DeloadThreshold: 3, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "increment",
		},
		{
			name: "linear negative increment",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &neg,
				DeloadThreshold: 3, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "increment",
		},
		{
			name: "percentage missing increment_pct",
			r: ProgressionRule{
				Strategy:        ProgressionPercentage,
				DeloadThreshold: 3, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "increment_pct",
		},
		{
			name: "deload_threshold zero",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &inc5,
				DeloadThreshold: 0, DeloadPct: 10,
			},
			wantErr:  true,
			errField: "deload_threshold",
		},
		{
			name: "deload_pct over 100",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &inc5,
				DeloadThreshold: 3, DeloadPct: 101,
			},
			wantErr:  true,
			errField: "deload_pct",
		},
		{
			name: "deload_pct negative",
			r: ProgressionRule{
				Strategy: ProgressionLinear, Increment: &inc5,
				DeloadThreshold: 3, DeloadPct: -1,
			},
			wantErr:  true,
			errField: "deload_pct",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.r.Validate()
			assertValidationResult(t, err, tt.wantErr, tt.errField)
		})
	}
}

func TestProgressionRule_NextWeight(t *testing.T) {
	inc := 5.0
	pct := 2.5

	tests := []struct {
		name     string
		rule     ProgressionRule
		current  float64
		failures int
		expected float64
	}{
		{
			name:     "linear increment",
			rule:     ProgressionRule{Strategy: ProgressionLinear, Increment: &inc, DeloadThreshold: 3, DeloadPct: 10},
			current:  100, failures: 0,
			expected: 105,
		},
		{
			name:     "percentage increment",
			rule:     ProgressionRule{Strategy: ProgressionPercentage, IncrementPct: &pct, DeloadThreshold: 3, DeloadPct: 10},
			current:  200, failures: 0,
			expected: 205,
		},
		{
			name:     "wave returns current",
			rule:     ProgressionRule{Strategy: ProgressionWave, DeloadThreshold: 3, DeloadPct: 10},
			current:  100, failures: 0,
			expected: 100,
		},
		{
			name:     "deload triggered",
			rule:     ProgressionRule{Strategy: ProgressionLinear, Increment: &inc, DeloadThreshold: 3, DeloadPct: 10},
			current:  100, failures: 3,
			expected: 90,
		},
		{
			name:     "deload at exact threshold",
			rule:     ProgressionRule{Strategy: ProgressionPercentage, IncrementPct: &pct, DeloadThreshold: 2, DeloadPct: 20},
			current:  200, failures: 2,
			expected: 160,
		},
		{
			name:     "hold on failure below threshold",
			rule:     ProgressionRule{Strategy: ProgressionLinear, Increment: &inc, DeloadThreshold: 3, DeloadPct: 10},
			current:  100, failures: 2,
			expected: 100,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.rule.NextWeight(tt.current, tt.failures)
			diff := got - tt.expected
			if diff < -0.001 || diff > 0.001 {
				t.Errorf("expected %.2f, got %.2f", tt.expected, got)
			}
		})
	}
}

File: app/domain/__tests__/progression.test.ts
import { nextWeight, validateProgressionRule } from "../progression";

describe("nextWeight", () => {
  const linearRule = {
    strategy: "linear" as const,
    increment: 2.5,
    increment_pct: null,
    deload_threshold: 3,
    deload_pct: 10,
  };

  const percentageRule = {
    strategy: "percentage" as const,
    increment: null,
    increment_pct: 5,
    deload_threshold: 3,
    deload_pct: 10,
  };

  const waveRule = {
    strategy: "wave" as const,
    increment: null,
    increment_pct: null,
    deload_threshold: 3,
    deload_pct: 10,
  };

  it("increments linearly on zero failures", () => {
    expect(nextWeight(linearRule, 100, 0)).toBe(102.5);
  });

  it("increments by percentage on zero failures", () => {
    expect(nextWeight(percentageRule, 100, 0)).toBe(105);
  });

  it("holds weight on wave strategy", () => {
    expect(nextWeight(waveRule, 100, 0)).toBe(100);
  });

  it("holds weight on 1 failure (below threshold)", () => {
    expect(nextWeight(linearRule, 100, 1)).toBe(100);
  });

  it("holds weight on 2 failures (below threshold of 3)", () => {
    expect(nextWeight(linearRule, 100, 2)).toBe(100);
  });

  it("deloads on failures equal to threshold", () => {
    // 100 * (1 - 10/100) = 90
    expect(nextWeight(linearRule, 100, 3)).toBe(90);
  });

  it("deloads on failures exceeding threshold", () => {
    expect(nextWeight(linearRule, 100, 5)).toBe(90);
  });

  it("deloads percentage rule the same way", () => {
    expect(nextWeight(percentageRule, 200, 3)).toBe(180);
  });

  it("deloads wave rule the same way", () => {
    expect(nextWeight(waveRule, 100, 3)).toBe(90);
  });

  it("handles zero current weight", () => {
    expect(nextWeight(linearRule, 0, 0)).toBe(2.5);
  });
});

describe("validateProgressionRule", () => {
  it("accepts valid linear rule", () => {
    expect(
      validateProgressionRule({
        strategy: "linear",
        increment: 2.5,
        deload_threshold: 3,
        deload_pct: 10,
      })
    ).toBeNull();
  });

  it("rejects linear without increment", () => {
    const err = validateProgressionRule({
      strategy: "linear",
      increment: null,
      deload_threshold: 3,
      deload_pct: 10,
    });
    expect(err).not.toBeNull();
    expect(err!.field).toBe("increment");
  });

  it("rejects invalid strategy", () => {
    const err = validateProgressionRule({
      strategy: "invalid",
      deload_threshold: 3,
      deload_pct: 10,
    });
    expect(err).not.toBeNull();
    expect(err!.field).toBe("strategy");
  });

  it("rejects deload_threshold < 1", () => {
    const err = validateProgressionRule({
      strategy: "linear",
      increment: 2.5,
      deload_threshold: 0,
      deload_pct: 10,
    });
    expect(err).not.toBeNull();
    expect(err!.field).toBe("deload_threshold");
  });

  it("rejects deload_pct > 100", () => {
    const err = validateProgressionRule({
      strategy: "linear",
      increment: 2.5,
      deload_threshold: 3,
      deload_pct: 150,
    });
    expect(err).not.toBeNull();
    expect(err!.field).toBe("deload_pct");
  });

  it("accepts valid percentage rule", () => {
    expect(
      validateProgressionRule({
        strategy: "percentage",
        increment_pct: 5,
        deload_threshold: 3,
        deload_pct: 10,
      })
    ).toBeNull();
  });

  it("rejects percentage without increment_pct", () => {
    const err = validateProgressionRule({
      strategy: "percentage",
      increment_pct: null,
      deload_threshold: 3,
      deload_pct: 10,
    });
    expect(err).not.toBeNull();
    expect(err!.field).toBe("increment_pct");
  });
});
